### PR TITLE
Dependency graph plotter

### LINKED ISSUE #69: Graph visualisation
Add function to render  all dependencies into one file. HTML with Mermaid can be good for start

### PR DESCRIPTION
**Pull request by Tishka17**
/closes #69

### CODE REVIEW

File: src/dishka/plotter/mermaid.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        )
+
+    def _render_node_deps(self, node: Node) -> list[str]:
+        return [
+            f"{dep} ..> {node.id}"
+            for dep in node.dependencies
+        ]
+
+    def _group_type(self, group: Group) -> str:
+        return ""
```
Comment (Tishka17) [inline src/dishka/plotter/mermaid.py:48]: add emoji

File: src/dishka/plotter/d2.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        self.names: dict[str, str] = {}
+        self.full_ids: dict[str, str] = {}
+
+    def _render_node(self, node: Node) -> str:
+        name = self._node_type(node) + node.name
+        if node.source_name:
+            source = f'"{node.source_name}()": ""\n'
+        else:
+            source = ""
+        node_id = node.id
```
Comment (IvanKirpichnikov) [inline src/dishka/plotter/d2.py:15]: an unnecessary variable

File: src/dishka/plotter/d2.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        elif node.type is NodeType.ALIAS:
+            return "🔗 " + prefix
+        return "🏭 " + prefix
+
+    def _render_group(
+            self, group: Group, name_prefix: str = "",
+    ) -> str:
+        name = self._group_type(group) + name_prefix + (group.name or "")
+        res = ""
+        res += f'{group.id}: "{name}"' + "{\n"
```
Comment (IvanKirpichnikov) [inline src/dishka/plotter/d2.py:53]: maybe `res = f'{group.id}: "{name}"' + "{\n"`?

File: src/dishka/plotter/mermaid.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        for group in groups:
+            for node in group.nodes:
+                self.names[node.id] = node.name
+            self._fill_names(group.children)
+
+    def render(self, groups: list[Group]) -> str:
+        self._fill_names(groups)
+
+        res = "classDiagram\n"
+        res += "direction LR\n"
```
Comment (IvanKirpichnikov) [inline src/dishka/plotter/mermaid.py:94]: meybe `res = "classDiagram\ndirection LR\n"`?

File: src/dishka/plotter/transform.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    def _is_protocol(self, key: DependencyKey) -> bool:
+        return Protocol in getattr(key.type_hint, "__bases__", [])
+
+    def _is_empty(self, registry: Registry) -> bool:
+        if len(registry.factories)>1:
+            return False
+        key = next(iter(registry.factories))
+        if key.type_hint in (Container, AsyncContainer):
+            return True
+        return False
```
Comment (IvanKirpichnikov) [inline src/dishka/plotter/transform.py:30]: `return key.type_hint in (Container, AsyncContainer)`?

File: src/dishka/plotter/transform.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        self.nodes: dict[tuple[DependencyKey, BaseScope], Node] = {}
+        self.groups: dict[Any, Group] = {}
+        self._counter = 0
+
+    def count(self, prefix: str) -> str:
+        self._counter += 1
+        return f"{prefix}{self._counter}"
+
+    def _is_protocol(self, key: DependencyKey) -> bool:
+        return Protocol in getattr(key.type_hint, "__bases__", [])
```
Comment (IvanKirpichnikov) [inline src/dishka/plotter/transform.py:22]: use the `is_protocol` function from the `_adaptix/type_tools` package

**Pull request merged**